src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Telemetry/TelemetryReporter.cs

```diff
@@ -22,22 +24,42 @@ public TelemetryReporter(ImmutableArray<TelemetrySession> telemetrySessions, ILo
 
         public void ReportEvent(string name, TelemetrySeverity severity)
         {
-            var telemetryEvent = new TelemetryEvent(name, severity);
+            var telemetryEvent = new TelemetryEvent(GetTelemetryName(name), severity);
             Report(telemetryEvent);
         }
 
+        public void ReportEvent(string name, TelemetrySeverity severity, ImmutableDictionary<string, object> values)
+        {
+            var telemetryEvent = new TelemetryEvent(GetTelemetryName(name), severity);
+            foreach (var kvp in values)
+            {
+                var propertyName = kvp.Key;
+                var propertyValue = kvp.Value;
+                telemetryEvent.Properties.Add(GetPropertyName(propertyName), new TelemetryComplexProperty(propertyValue));
+            }
```

Comment:
Hmmm... should we add a `Deconstruct` extension method for `KeyValuePair<,>` in `Microsoft.AspNetCore.Razor.Common` so that you can use tuple deconstruction syntax when foreach-ing a dictionary? Something like this?

```C#
public static void Deconstruct<TKey, TValue>(this KeyValuePair<TKey, TValue> kvp, out TKey key, out TValue value)
    => (key, value) = (kvp.Key, kvp.Value);
```

Reply:
Love it. Wasn't sure what the plan was, but having deconstruct would be valuable. Will udpate